src/js/client/editprefset-dialogs-widget.js

```diff
@@ -0,0 +1,348 @@
+    templates: {
+        initial: "editprefset-confirmAddProduct-dialog"
+    },
+    model: {
+        appId: null,
+        name: null, // Human Name of Product,
+        allSolutions: null,
+        contextNames: null
+    },
+    invokers: {
+        acceptConfirmDialog: {
+            funcName: "gpii.devpmt.dialogs.confirmAddProductDialog.acceptConfirmDialog",
+            args: ["{that}.closeDialog", "{that}.model.appId", "{gpii.devpmt.editPrefs}.model.allSolutions", "{gpii.devpmt.editPrefs}.model.contextNames", "{gpii.devpmt.editPrefs}.editProductEnabled"]
+        }
+    }
+});
+
+/**
+ * Clicking Add on the Product Dialog adds this product to the preference set.
+ * A preference safe must have at least one context (prefset) in order to add
+ * a product to it.
+ *
+ * @param {Function} closeDialog - Zero-arg function to close the dialog window.
+ * @param {Object} that - Dialog instance
+ * @param {String} appId - String indicating the appId
+ *                                     ex. http://registry.gpii.net/applications/com.android.freespeech
+ * @param {Object} solutions - Solutions registry entries.
+ * @param {Array} contextNames - Array of Strings with the name/id the context is
+ * keyed by. The product will be initially added to the first item in this array.
+ * @param {Function} editProductEnabled - Invoker from `gpii.devpmt.editPrefs` to add the
+ * product to.
+ */
+gpii.devpmt.dialogs.confirmAddProductDialog.acceptConfirmDialog = function (closeDialog, appId, allSolutions, contextNames, editProductEnabled) {
+    // In this case we actually need to close the dialog first... as the page
+    // rerenders based on a model listener when the product is enabled, and does
+    // wonky things... such as removing the ability to vertical scroll. Should look
+    // at reworking this perhaps.
+    closeDialog();
+    // It's unlikely that this dialog could have been instantiated with an appId not from
+    // the list, but in any event, if the appId is not in the solutions listing, we will
+    // return here.
+    if (!allSolutions[appId]) {
+        return;
+    }
+    // TODO Ontology!!!
+    var appUrl = "http://registry.gpii.net/applications/" + appId;
+    if (contextNames.length > 0) {
+        editProductEnabled(true, contextNames[0],  appUrl);
+    }
+};
+
+/**
+ * Confirm Preferences Set Save Dialog
+ */
+fluid.defaults("gpii.devpmt.dialogs.confirmSaveDialog", {
+    gradeNames: ["gpii.devpmt.dialogs.confirmDialog"],
+    templates: {
+        initial: "editprefset-confirmSave-dialog"
+    },
+    model: {
+        unsavedChanges: []
+    },
+    invokers: {
+        acceptConfirmDialog: {
+            funcName: "gpii.devpmt.dialogs.confirmSaveDialog.acceptConfirmDialog",
+            args: ["{that}.closeDialog", "{gpii.devpmt.editPrefs}.savePrefset"]
+        }
+    }
+});
+
+gpii.devpmt.dialogs.confirmSaveDialog.acceptConfirmDialog = function (closeDialog, saveFunc) {
+    saveFunc();
```

Comment:
This pattern is a sign that what you have here is really an event rather than an invoker.
Instead shift acceptConfirmDialog into "events" and then write these entries as two namespaced listeners, with, e.g.
listeners: {
    "acceptConfirmDialog.closeDialog": {
        priority: "after:save",
        func: "{gpii.devpmt.editPrefs}.savePrefset"
    }

This leads to the path to openness.

Reply:
Thanks!  That was a good idea, and I thoroughly enjoyed that little bit of refactoring.